catalog/registry.go

```diff
@@ -0,0 +1,135 @@
+	r[catalogType] = reg
+}
+
+func (r registry) get(catalogType string) (Registrar, bool) {
+	regMutex.Lock()
+	defer regMutex.Unlock()
+	reg, ok := r[catalogType]
+	return reg, ok
+}
+
+func (r registry) remove(catalogType string) {
+	regMutex.Lock()
+	defer regMutex.Unlock()
+	delete(r, catalogType)
+}
+
+var (
+	regMutex        sync.Mutex
+	defaultRegistry = registry{}
+)
+
+// Registrar is a factory for creating Catalog instances, used for registering to use
+// with LoadCatalog.
+type Registrar interface {
+	GetCatalog(catalogURI string, props iceberg.Properties) (Catalog, error)
+}
+
+type RegistrarFunc func(string, iceberg.Properties) (Catalog, error)
+
+func (f RegistrarFunc) GetCatalog(catalogURI string, props iceberg.Properties) (Catalog, error) {
+	return f(catalogURI, props)
+}
+
+// Register adds the new catalog type to the registry. If the catalog type is already registered, it will be replaced.
+func Register(catalogType string, reg Registrar) {
+	if reg == nil {
+		panic("catalog: RegisterCatalog catalog factory is nil")
+	}
+	defaultRegistry.set(catalogType, reg)
+}
+
+// Unregister removes the requested catalog factory from the registry.
+func Unregister(catalogType string) {
+	defaultRegistry.remove(catalogType)
+}
+
+// GetRegsisteredCatalogs returns the list of registered catalog names that can
+// be looked up via LoadCatalog.
+func GetRegisteredCatalogs() []string {
+	return defaultRegistry.getKeys()
+}
+
+// Load allows loading a specific catalog by URI and properties.
+//
+// This is utilized alongside RegisterCatalog/UnregisterCatalog to not only allow
+// easier catalog loading but also to allow for custom catalog implementations to
+// be registered and loaded external to this module.
+//
+// The URI is used to determine the catalog type by first checking if it contains
+// the string "://" indicating the presence of a scheme. If so, the schema is used
+// to lookup the registered catalog. i.e. "glue://..." would return the Glue catalog
+// implementation, passing the URI and properties to NewGlueCatalog. If no scheme is
+// present, then the URI is used as-is to lookup the catalog factory function.
+//
+// Currently the following catalogs are registered by default:
+//
+//   - "glue" for AWS Glue Data Catalog, the rest of the URI is ignored, all configuration
+//     should be provided using the properties. "glue.region", "glue.endpoint",
+//     "glue.max-retries", etc. Default AWS credentials are used if found, or can be
+//     overridden by setting "glue.access-key", "glue.secret-access-key", and "glue.session-token".
+//
+//   - "rest" for a REST API catalog, if the properties have a "uri" key, then that will be used
+//     as the REST endpoint, otherwise the URI is used as the endpoint. The REST catalog also
+//     registers "http" and "https" so that Load with an http/s URI will automatically
+//     load the REST Catalog.
+func Load(catalogURI string, props iceberg.Properties) (Catalog, error) {
```

Comment:
Do we want to support connecting to multiple of the same catalog types? 

pyiceberg's [`load_catalog`](https://py.iceberg.apache.org/reference/pyiceberg/catalog/#pyiceberg.catalog.load_catalog) references catalog by name. And the URI is part of properties dict https://py.iceberg.apache.org/configuration/#rest-catalog

Same for [CatalogUtil.loadCatalog](https://github.com/apache/iceberg/blob/67e084c8d47db68c6135f9837b9cff6f88da0309/core/src/main/java/org/apache/iceberg/CatalogUtil.java#L251-L276) and [FlinkCatalogFactory](https://github.com/apache/iceberg/blob/main/flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java#L88-L123)  in java

Reply:
in the example, the catalog name can be the catalog uri, but we should make this distinction in the functions

Reply:
That is supported in this structure (unless I'm missing something) calling `Load` will ultimately call `GetCatalog` on the Registered factory, creating a *new* catalog instance each time you call `catalog.Load`

i.e.: you can call `catalog.Load("rest", iceberg.Properties{"uri": "http://foobar"})`  and `catalog.Load("http://foobarbaz", nil)` and you'll get two *different* rest catalog instances each using their appropriate endpoint. 

The distinction is that we allow simplification by checking the scheme of the URI instead of always having a separate "name"

Reply:
Do we want to support some kind of caching in the `Registrar`?
Right now, calling `Load("http://my-rest"))` twice will creating 2 separate connections to the backend

Reply:
that makes sense, thanks for the explanation!